modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/IgniteStatisticsRepositoryImpl.java

```diff
@@ -0,0 +1,245 @@
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ignite.internal.processors.query;
+
+import org.apache.ignite.IgniteLogger;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.cache.GridCacheUtils;
+import org.apache.ignite.internal.processors.cache.query.QueryTable;
+import org.apache.ignite.internal.processors.query.stat.ColumnStatistics;
+import org.apache.ignite.internal.processors.query.stat.IgniteStatisticsManagerImpl;
+import org.apache.ignite.internal.processors.query.stat.IgniteStatisticsRepository;
+import org.apache.ignite.internal.processors.query.stat.ObjectPartitionStatistics;
+import org.apache.ignite.internal.processors.query.stat.ObjectStatistics;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class IgniteStatisticsRepositoryImpl implements IgniteStatisticsRepository {
+    /** Logger. */
+    private IgniteLogger log;
+
+    /** */
+    private final GridKernalContext ctx;
+
+    /** Table->Partition->Partition Statistics map, populated only on server nodes without persistence enabled.  */
+    private final Map<QueryTable, Map<Integer, ObjectPartitionStatistics>> partsStats;
+
+    /** Local (for current node) object statistics. */
+    private final Map<QueryTable, ObjectStatistics> localStats;
+
+    /** Global (for whole cluster) object statistics. */
+    private final Map<QueryTable, ObjectStatistics> globalStats = new ConcurrentHashMap<>();
+
+
+    public IgniteStatisticsRepositoryImpl(GridKernalContext ctx) {
+        this.ctx = ctx;
+
+        if (ctx.config().isClientMode() || ctx.isDaemon()) {
+            // Cache only global statistics, no store
+            partsStats = null;
+            localStats = null;
+        } else {
+            if (GridCacheUtils.isPersistenceEnabled(ctx.config())) {
+                // Persistence store
+                partsStats = null;
+            } else {
+                // Cache partitions statistics, no store
+                partsStats = new ConcurrentHashMap<>();
+            }
+            localStats = new ConcurrentHashMap<>();
+        }
+        log = ctx.log(IgniteStatisticsRepositoryImpl.class);
+    }
+
+    @Override public void saveLocalPartitionsStatistics(QueryTable tbl, Collection<ObjectPartitionStatistics> statistics,
+                                                        boolean fullStat) {
+        if (partsStats != null) {
```

Comment:
it's possible to decrease nesting level by inverting the condition

Reply:
There will be some problems when we'll add storage. Because here we should cache statistics in partsStats or store it into metastore.

Reply:
and do you have strong vision about how it will be?